File: solution files/Python/part-03/task-01-refactor/blueprint_decoder.py
import re
from typing import Dict, List


def read_blueprint(filename: str) -> str:
    """Read and return the raw content of a blueprint file."""
    with open(filename, "r") as f:
        return f.read()


def extract_secrets(content: str) -> List[str]:
    """Extract all secrets marked between {* and *} from content."""
    return re.findall(r"\{\* (.*?) \*\}", content)


def categorize_secrets(secrets: List[str]) -> Dict[str, int]:
    """Count secrets by their category prefix (text before the first colon)."""
    categories: Dict[str, int] = {}
    for secret in secrets:
        key = secret.split(":")[0].strip() if ":" in secret else "UNKNOWN"
        categories[key] = categories.get(key, 0) + 1
    return categories


def decode_blueprint_safe(filename: str) -> List[str]:
    """Full pipeline: read file safely and extract secrets."""
    try:
        content = read_blueprint(filename)
        return extract_secrets(content)
    except FileNotFoundError:
        print(f"Error: '{filename}' not found.")
        return []

File: solution files/C#/part-03/task-01-refactor/BlueprintDecoder.cs
using System.IO;
using System.Text.RegularExpressions;
using System.Collections.Generic;
using System.Linq;

public class BlueprintDecoder
{
    private static readonly Regex SecretPattern = new Regex(@"\{\* (.*?) \*\}", RegexOptions.Compiled);

    /// <summary>Read and return the raw content of a blueprint file.</summary>
    public static string ReadBlueprint(string filename) => File.ReadAllText(filename);

    /// <summary>Extract all secrets marked between {* and *} from content.</summary>
    public static List<string> ExtractSecrets(string content) =>
        SecretPattern.Matches(content)
                     .Select(m => m.Groups[1].Value)
                     .ToList();

    /// <summary>Count secrets by their category prefix (text before the first colon).</summary>
    public static Dictionary<string, int> CategorizeSecrets(List<string> secrets) =>
        secrets.GroupBy(s => s.Contains(':') ? s.Split(':')[0].Trim() : "UNKNOWN")
               .ToDictionary(g => g.Key, g => g.Count());

    /// <summary>Full pipeline: read file safely and extract secrets.</summary>
    public static List<string> DecodeBlueprintSafe(string filename)
    {
        try { return ExtractSecrets(ReadBlueprint(filename)); }
        catch (FileNotFoundException) { System.Console.WriteLine($"Error: '{filename}' not found."); return new(); }
    }
}
